dnssec-tools/validator/doc/gethost.1
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.35
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "GETHOST 1"
.TH GETHOST 1 "2011-06-28" "perl v5.8.9" "User Commands"
.SH "NAME"
gethost \- command\-line test program for the \fIval_gethostbyname()\fR (and related) 
functions 
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\&   gethost [options] name
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
This utility is a command-line wrapper around the \fI\fIval_gethostbyname()\fI\fR
(and related) functions.  It invokes the \fI\fIval_gethostbyname()\fI\fR,
\&\fI\fIval_gethostbyname_r()\fI\fR and \fI\fIval_gethostbyname2()\fI\fR functions for the given
command-line arguments and displays the returned \fIhostent\fR, the validation
status value, and the value of the \fIh_errno\fR variable.
.PP
The exit status for this program is 1 if the answer received is trusted, 2 if
the answer received is validated, and \-1 for an error.  The program returns 0 if 
no validation was performed or if the result was neither trusted nor validated. 
The trusted and validated status values are obtained using the 
\&\fI\fIval_istrusted()\fI\fR and \fI\fIval_isvalidated()\fI\fR functions from \fI\fIlibval\fI\|(3)\fR.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\-r, \-\-reentrant" 4
.IX Item "-r, --reentrant"
Use the re-entrant query interface.
.IP "\-f, \-\-family=\fI[AF_INET|AF_INET6]\fR" 4
.IX Item "-f, --family=[AF_INET|AF_INET6]"
Use the specified address family for the query. 
.IP "\-o, \-\-output=<debug\-level>:<dest\-type>[:<dest\-options>]" 4
.IX Item "-o, --output=<debug-level>:<dest-type>[:<dest-options>]"
<debug\-level> is 1\-7, corresponding to syslog levels ALERT-DEBUG
<dest\-type> is one of file, net, syslog, stderr, stdout
<dest\-options> depends on <dest\-type>
    file:<file\-name>   (opened in append mode)
    net[:<host\-name>:<host\-port>] (127.0.0.1:1053
    syslog[:facility] (0\-23 (default 1 \s-1USER\s0))
.IP "\-h, \-\-help" 4
.IX Item "-h, --help"
Display the help and exit.
.SH "PRE-REQUISITES"
.IX Header "PRE-REQUISITES"
libval
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright 2005\-2011 \s-1SPARTA\s0, Inc.  All rights reserved.
See the \s-1COPYING\s0 file included with the DNSSEC-Tools package for details.
.SH "AUTHORS"
.IX Header "AUTHORS"
Abhijit Hayatnagarkar, Suresh Krishnaswamy, Robert Story
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fB\f(BIval_getaddrinfo\fB\|(3)\fR
.PP
\&\fB\f(BIlibval\fB\|(3)\fR
.PP
http://dnssec\-tools.sourceforge.net


dnssec-tools/NEWS
This file contains a list of major changes per release.  See the
ChangeLog file for a complete set of changes and their details
details.


1.10
 - New Features:
    - New Apps:     (see the validator/apps directory for details)
                    - dnssec-check: check dnssec support from your ISP
                    - dnssec-nodes: graphically displays a DNS
		      hierarchy, color coded by each node's DNSSEC status
                    - dnssec-system-tray: displays pop-up
     		      notifications when a libval-enabled application
		      triggers a DNSSEC error
                    - lookup: a graphical DNS lookup utility that
		      displays the results in a hierarchical tree and
		      color codes the window according to DNSSEC status

    - libval:       - Added support for building on Windows.
                    - added support for falling back to recursion when
                      the caching name server does not appear to 
                      support DNSSEC. This also works as a mechanism
                      to work around poisoned or misbehaving cache. 
                    - Significant improvements to the the asynchronous support. 
    - lsdnssec:     - Improvements to lsdnssec to display different
                      output depending on whether a zone is a
                      stand-alone zone or under control of rollerd.
    - nagios:       - Plugins for the nagios monitoring system which
                      enable monitoring of zone rollerover states.
    - firefox:      - Improved patches that work with the most recent firefox

 Plus many more minor features and bug fixes

1.9
 - New Features:
    - lsdnssec:     - Added a new flag (-p) to show only zones in a
                      particular rollerd phase.
                    - fixed bugs to align timing output with rollerd.
    - rollerd:      - Added a -logtz flag for logging timezones
                    - fixed bugs related to the -alwayssign flag.
		    - zonesigner's path is taken from the config file.
    - rollctl:      - Added -rollall and -rollzone options.
    - zonesigner:   - Assumes keys need to be generated for new zones
		      (Assumes -genkeys option was given if a keyrec file
		      can't be found.)
                    - Exits with unique exit codes if a failure occurs.
                      ("zonesigner -xc CODE" can lookup a description for it.")
		    - Added the -phase option so rollover options could be
		      more easily specified.
    - lights:       - A simple GUI to check the status of rollover states
    - blinkenlights:- Added hide/show commands for rollrec names and zone
		      names, for split-zone support
    - cleankrf:	    - Fixed deletion of obsolete set keyrecs.
    - GUI commands: - Fixed how the Exit command works so they don't coredump.

    - libsres
      & libval:     - New beta support for issuing asynchronous requests.
                      This can speed up queries by up to 4 times if used.
                      (see example code in validator/apps/validator_selftest.c)
                    - NSEC3, DLV and IPv6 are enabled by default.
                    - improved logging and logging-callback support.
    - drawvalmap    - Can output PNG files now

 - Packaging:
                    - Our download page now allows you to download
                      the C validator libraries independently of the
                      full DNNSEC-Tools tool-suite.

 - Many bugs were also fixed in the 240+ changes.

1.8
 - New Features:
    - zonesigner, rollerd
                    - Made changes so that these tools are more compatible 
                      with recent versions of Bind
                    - The zone_errors configuration parameter allows a zone-
                      specific maximum to be set.  Once exceeded, that zone
                      will be skipped rather than allowing rollover to continue.
    - blinkenlights
                    - Recognizes when rollerd abruptly quits, so error messages
                      aren't spewed interminably.
    - ZonFile::Fast - Fixed parsing of DS records containing spaces and 
                      parsing of mname and rname SOA fields
                    - Added support for parsing KEY records
    - keyrec.pm     - Made changes to properly lock keyrec files before 
                      writing to them.
                    - Begun process of deprecating keyrec_open().
    - mapper:       - added a new option: --node-size for mapping
                      complex zones.
    - dnspktflow:   - added two new options:
                      --layout-style for selecting the layout style to use
                      --node-size for mapping complex zones.
                    - Add new (default) option to cluster
                      authoritative nodes together to help better
                      understand the relationships between traffic
                      patterns and authoritative name server/zone arrangement.
    - libval:       - Now distributed with the Root TA.
                    - Added stricter checks for openssl SHA-256 support in 
                      configure. 
                    - Added several improvements that allow the validator to 
                      lookup information within provably insecure zones that 
                      do not handle EDNS0 requests nicely. This includes 
                      adding support for turning off EDNS0 when traversing a 
                      name hierarchy that leads to a provably insecure zone, 
                      EDNS0 fallback support, and additional checks to check 
                      the sanity of response data.
                    - Fixed certain bugs in CNAME handling and in the 
                      validation of proofs accompanying wildcard responses, 
                      referrals and alias chains.
                    - Fixed support for RSADSA and RSASHA-512 signature 
                      validation.

dnssec-tools/validator/apps/gethost.c
/*
 * Copyright 2005-2011 SPARTA, Inc.  All rights reserved.
 * See the COPYING file distributed with this software for details.
 *
 * Author: Abhijit Hayatnagarkar
 *
 * A command-line tool for testing the val_gethostbyname*() functions.
 */

#include "validator-config.h"
#include <validator/validator.h>

#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif

#define	NAME	"gethost"
#define	VERS	"version: 1.0"
#define	DTVERS	"DNSSEC-Tools Version: 1.8"

#ifdef HAVE_GETOPT_LONG
// Program options
static struct option prog_options[] = {
    {"help", 0, 0, 'h'},
    {"family", 0, 0, 'f'},
    {"reentrant", 0, 0, 'r'},
    {"output", 0, 0, 'o'},
    {"Version", 0, 0, 'V'},
    {0, 0, 0, 0}
};
#endif

#define AUX_BUFLEN 16000

void
usage(char *progname)
{
    /* *INDENT-OFF* */
    fprintf(stderr, "Usage: %s [options] name\n", progname);
    fprintf(stderr, "Options:\n");
    fprintf(stderr, "\t-h, --help                      display usage and exit\n");
    fprintf(stderr, "\t-r, --reentrant                 use reentrant versions of functions\n");
    fprintf(stderr, "\t-f, --family=[AF_INET|AF_INET6] address family\n");
    fprintf(stderr, "\t                                AF_INET for IPv4 addresses,\n");
    fprintf(stderr, "\t                                and AF_INET6 for IPv6 addresses\n");
    fprintf(stderr,
            "\t-o, --output=<debug-level>:<dest-type>[:<dest-options>]\n"
            "\t          <debug-level> is 1-7, corresponding to syslog levels\n"
            "\t          <dest-type> is one of file, net, syslog, stderr, stdout\n"
            "\t          <dest-options> depends on <dest-type>\n"
            "\t              file:<file-name>   (opened in append mode)\n" 
            "\t              net[:<host-name>:<host-port>] (127.0.0.1:1053\n" 
            "\t              syslog[:facility] (0-23 (default 1 USER))\n" );
    /* *INDENT-ON* */
    fprintf(stderr, "\t-V, --Version                   display version and exit\n");
}

void
version(void)
{
    fprintf(stderr, "%s: %s\n", NAME,VERS);
    fprintf(stderr, "%s\n", DTVERS);
}

int
main(int argc, char *argv[])
{
    struct hostent  hentry;
    char            auxbuf[AUX_BUFLEN];
    struct hostent *result = NULL;

    int             i;
    val_status_t    val_status;
    int             herrno = 0;
    int             familyspecified = 0;
    int             usereentrant = 0;
    char           *name;
    int             af = AF_INET;
    char            buf[INET6_ADDRSTRLEN];
    size_t          buflen = INET6_ADDRSTRLEN;
    int             retval = 0;
    val_log_t  *logp;

    memset(&hentry, 0, sizeof(struct hostent));
    memset(auxbuf, 0, AUX_BUFLEN);

    // Parse the command line

    while (1) {
        int             c;
#ifdef HAVE_GETOPT_LONG
        int             opt_index = 0;
#ifdef HAVE_GETOPT_LONG_ONLY
        c = getopt_long_only(argc, argv, "hrf:o:V",
                             prog_options, &opt_index);
#else
        c = getopt_long(argc, argv, "hrf:o:V", prog_options, &opt_index);
#endif
#else                           /* only have getopt */
        c = getopt(argc, argv, "hrf:o:V");
#endif

        if (c == -1) {
            break;
        }

        switch (c) {
        case 'h':
            usage(argv[0]);
            return -1;
        case 'f':
            familyspecified = 1;
            if (strncasecmp(optarg, "AF_INET", strlen("AF_INET")) == 0) {
                af = AF_INET;
            } else if (strncasecmp(optarg, "AF_INET6", strlen("AF_INET6")) == 0) {
                af = AF_INET6;
            } else {
                fprintf(stderr, "Invalid family %s\n", optarg);
                usage(argv[0]);
                return -1;
            }
            break;
        case 'r':
            usereentrant = 1;
            break;
        case 'o':
            logp = val_log_add_optarg(optarg, 1);
            if (NULL == logp) { /* err msg already logged */
                usage(argv[0]);
                return -1;
            }
            break;
        case 'V':
            version();
            return 0;

        default:
            fprintf(stderr, "Invalid option %s\n", argv[optind - 1]);
            usage(argv[0]);
            return -1;
        }
    }


    if (optind < argc) {
        name = argv[optind++];
    } else {
        fprintf(stderr, "Error: name not specified\n");
        usage(argv[0]);
        return -1;
    }

    if (usereentrant) {
#ifdef HAVE_GETHOSTBYNAME2
            if (familyspecified)
                retval =
                    val_gethostbyname2_r(NULL, name, af, &hentry, auxbuf,
                                         AUX_BUFLEN, &result, &herrno,
                                         &val_status);
            else
#endif
                retval =
                    val_gethostbyname_r(NULL, name, &hentry, auxbuf,
                                        AUX_BUFLEN, &result, &herrno,
                                        &val_status);
    } else {
            if (familyspecified)
                result = val_gethostbyname2(NULL, name, af, &val_status);
            else
                result = val_gethostbyname(NULL, name, &val_status);
    }

    if (result != NULL) {
        const char *addr = NULL;

        printf("\n\th_name = %s\n", result->h_name);
        if (result->h_aliases) {
            printf("\th_aliases = \n");
            for (i = 0; result->h_aliases[i] != 0; i++) {
                printf("\t\t[%d] = %s\n", i, result->h_aliases[i]);
            }
        } else
            printf("\th_aliases = NULL\n");
        printf("\th_length = %d\n", result->h_length);
        printf("\th_addr_list = \n");
        if (result->h_addrtype == AF_INET) {
            struct sockaddr_in sa;
            printf("\th_addrtype = AF_INET\n");
            for (i = 0; result->h_addr_list[i] != 0; i++) {
                memset(buf, 0, buflen);
                memcpy(&sa.sin_addr, result->h_addr_list[i],
                       sizeof(sa.sin_addr));
                INET_NTOP(AF_INET, ((struct sockaddr *)&sa), sizeof(sa), 
                    buf, buflen, addr);
                printf("\t\t[%d] = %s\n", i, addr);
            }
        } 
#ifdef VAL_IPV6
        else if (result->h_addrtype == AF_INET6) {
            struct sockaddr_in6 sa6;
            printf("\th_addrtype = AF_INET6\n");
            for (i = 0; result->h_addr_list[i] != 0; i++) {
                memset(buf, 0, buflen);
                memcpy(&sa6.sin6_addr, result->h_addr_list[i],
                       sizeof(sa6.sin6_addr));
                INET_NTOP(AF_INET6, ((struct sockaddr *)&sa6), sizeof(sa6), 
                    buf, buflen, addr);
                printf("\t\t[%d] = %s\n", i, addr);
            }
        } 
#endif
        else {
            printf("\th_addrtype = %d\n", result->h_addrtype);
        }
    } else {
        printf("result is NULL\n");
    }
    printf("Validation status = %s\n", p_val_error(val_status));
    if (usereentrant) {
#ifdef HAVE_HSTRERROR
        printf("h_errno = %s\n", hstrerror(herrno));
#else
        printf("h_errno = %d\n", herrno);
#endif
    } else {
#ifdef HAVE_HSTRERROR
        printf("h_errno = %s\n", hstrerror(h_errno));
#else
        printf("h_errno = %d\n", h_errno);
#endif
    }

    if (val_isvalidated(val_status))
        return 2;
    if (val_istrusted(val_status))
        return 1;

    return 0;
}
